src/graph_gui.cpp

```diff
@@ -347,36 +347,37 @@ struct BaseGraphWindow : Window {
 		/* Draw the background of the graph itself. */
 		GfxFillRect(r.left, r.top, r.right, r.bottom, GRAPH_BASE_COLOUR);
 
-		/* Draw the vertical grid lines. */
+		/* Draw the grid lines. */
+		int gridline_width = WidgetDimensions::scaled.bevel.top;
+		int grid_colour = GRAPH_GRID_COLOUR;
 
-		/* Don't draw the first line, as that's where the axis will be. */
+		/* Draw the vertical grid lines, skipping the first as that's where the axis will be. */
 		x = r.left + x_sep;
 
-		int grid_colour = GRAPH_GRID_COLOUR;
 		for (int i = 1; i < this->num_vert_lines + 1; i++) {
 			/* If using wallclock units, we separate periods with a lighter line. */
 			if (TimerGameEconomy::UsingWallclockUnits()) {
 				grid_colour = (i % 4 == 0) ? GRAPH_YEAR_LINE_COLOUR : GRAPH_GRID_COLOUR;
 			}
-			GfxFillRect(x, r.top, x, r.bottom, grid_colour);
+			GfxFillRect(x, r.top, x + gridline_width - 1, r.bottom, grid_colour);
 			x += x_sep;
 		}
 
 		/* Draw the horizontal grid lines. */
 		y = r.bottom;
 
 		for (int i = 0; i < (num_hori_lines + 1); i++) {
-			GfxFillRect(r.left - ScaleGUITrad(3), y, r.left - 1, y, GRAPH_AXIS_LINE_COLOUR);
-			GfxFillRect(r.left, y, r.right, y, GRAPH_GRID_COLOUR);
+			GfxFillRect(r.left - ScaleGUITrad(3), y, r.left - 1, y + gridline_width - 1, GRAPH_AXIS_LINE_COLOUR);
+			GfxFillRect(r.left, y, r.right + gridline_width - 1, y + gridline_width - 1, GRAPH_GRID_COLOUR);
 			y -= y_sep;
 		}
 
 		/* Draw the y axis. */
-		GfxFillRect(r.left, r.top, r.left, r.bottom, GRAPH_AXIS_LINE_COLOUR);
+		GfxFillRect(r.left, r.top, r.left + gridline_width - 1, r.bottom + gridline_width - 1, GRAPH_AXIS_LINE_COLOUR);
 
 		/* Draw the x axis. */
 		y = x_axis_offset + r.top;
-		GfxFillRect(r.left, y, r.right, y, GRAPH_ZERO_LINE_COLOUR);
+		GfxFillRect(r.left, y, r.right + gridline_width - 1, y + gridline_width - 1, GRAPH_ZERO_LINE_COLOUR);
```

Comment:
Not sure the right extent of the horizontal grid line should be extended. It's only the thickness of the line that needs to be adjusted.

Same for the y-axis, although there might be an exception if the gridlines meet at the bottom left corner.

Similar situations for the other lines, depending on horizontal vs vertical.

Reply:
I have thought about this and I think it is needed. It took me quite a few iterations to get all of these line extents pixel-perfect, and stems from the behaviour that `GfxFillRect(x, y, x, y)` draws 1x1px rather than nothing.

The horizontal lines are drawn over the vertical, and to draw the horizontal lines over the _final_ vertical line currently uses `r.right`. This gives a line 1px longer than you might expect (`length = r.right - r.left + 1`), and that extra 1px covers a 1px-wide final vertical line.

Based on that logic, to preserve the current behaviour with `gridline_width = 1`, `r.right + gridline_width - 1 = r.right + 1 - 1 = r.right`. So, mine is the naive correct solution for larger line widths. And, similar logic applies to the vertical extent of vertical lines.

(In this specific case, leaving it as `r.right` would partially cover the vertical line for wider lines. At, say, 4x GUI scale a 1px cap poking into the 4px wide line. The alternative is `r.right - 1`, which would remove the overlap of the final vertical line.)